### ISSUE TITLE
v2: Exceptions when accessing Country data where the value is null

### ISSUE THREAD

**Issue by limenet** [commerceguys/addressing#196]
Thanks for creating and maintaining this package, it's a huge help and #96 is appreciated!

---

After upgrading to v2, I noticed an issue when iterating over all countries (e.g. to seed a database). When executing the following snippet, an exception is thrown.

```php
use CommerceGuys\Addressing\Country\Country;
use CommerceGuys\Addressing\Country\CountryRepository;

$countryRepository = new CountryRepository();
$countries = $countryRepository->getAll();

collect($countries )
    ->map(fn (Country $country) => [
        'iso2' => $country->getCountryCode(),
        'iso3' => $country->getThreeLetterCode(),
    ]);
```

```
Typed property CommerceGuys\Addressing\Country\Country::$threeLetterCode must not be accessed before initialization at vendor/commerceguys/addressing/src/Country/Country.php:92
```

This is caused by `CountryRepository::getBaseDefinitions()` where entries `EA` and `IC` have `null` for the three-letter code (*). In `Country::__construct()`, the initialization of `$this->threeLetterCode` only occurs if `isset($definition['three_letter_code'])` and since `isset(null) === false`, this initialization never happens.

This behavior was introduced in d1c90a5ad5d0332c92c30b9d9f072b4eaf0352c0 by typing the property and [typed properties default to being uninitialized starting with PHP 7.4](https://php.watch/versions/7.4/typed-properties#uninitialized).

Unfortunately -- which is also why I didn't (yet) create a PR -- changing this behavior isn't quite as straightforward; either `protected string $threeLetterCode` would need to be changed to `protected ?string $threeLetterCode` (which might be considered a breaking change) or `Country::hasThreeLetterCode(): bool` would need to be added.

(*) Unverified, but this issue most likely also applies to the numeric code as well as the currency code for other entries such as `AQ`.

**Comment by bojanz** [comment]
Well I feel embarrassed :)

We caught this in beta2 for most classes, but apparently missed Country.

Pushed a fix. This change is fine cause we already document the matching getters as ?string.

Give it a whirl, and I'll tag v2.0.1 in the morning.

**Comment by limenet** [comment]
No worries :)

Thank you very much for the quick fix and I can confirm it's working as expected!

**Comment by limenet** [comment]
@bojanz sorry to bother you again! It appears 2.0.1 doesn't contain the fix, does it? There are no changes between 2.0.0 and 2.0.1 https://github.com/commerceguys/addressing/compare/v2.0.0...v2.0.1 and instead 54c70e7fb6146acbcceb5c937dd6cf9d13563bbe is after 2.0.1 https://github.com/commerceguys/addressing/compare/v2.0.1...master.

**Comment by bojanz** [comment]
The joys of developing on two laptops. I've repushed the release, but unsure if Packagist will know to clear its caches. If not, we'll do another tag.
